### DISCUSSION TITLE
Support for Hebrew

### DISCUSSION THREAD

**Discussion by Hilla-Merhav** [explosion/spaCy#12497] [Language Support]
Hello,
I am interested in adding full support for Hebrew in spaCy. After searching through the spaCy Discussions forum, I found [an older thread](https://github.com/explosion/spaCy/discussions/8234) discussing this same issue. 
Recently [a new UD corpus for Hebrew](https://github.com/UniversalDependencies/UD_Hebrew-IAHLTwiki) has been published and is licensed under CC BY-SA 4.0. However, I'm uncertain if this corpus can be used with spaCy due to license incompatibility. Can someone provide clarification on this matter?

**Comment by justus-saul** [comment]
Most of the models are trained on UD corpora and most of them have `CC BY-SA 4.0`. So it should be no issue.
Afaik license problems like "should not be used commercially" would be more a problem for the person wanting to use that model and not for Spacy.

**Comment by adrianeboyd** [comment]
Here are the slightly out-of-date basics for adding a new language to the trained pipelines: https://github.com/explosion/spaCy/discussions/3056

We look for corpora with licenses that allow commercial use that include annotations for: POS tags, lemmas, dependency parses, and named entities. Most languages use a UD corpus for POS+lemmas+dependencies and a separate NER corpus, but sometimes there's one corpus with all annotation layers. Typically NER corpora are more difficult to find with appropriate licenses, and Hebrew would probably need a third-party lemmatizer.

Hebrew in particular may be difficult because of spacy's limited ability to handle multi-word tokens from UD corpora and because spacy's built-in components do not perform well on Hebrew, see e.g. https://aclanthology.org/D19-3044.pdf.

**Reply by Hilla-Merhav** [reply]
Thank you for taking the time to answer my question. I was wondering if you could clarify whether [this UD corpus](https://github.com/UniversalDependencies/UD_Hebrew-IAHLTwiki), licensed under `CC BY-SA 4.0`, can be used.
Also, could you explain what "a third-party lemmatizer" means and why it might be necessary?
Thank you again for your help.

**Reply by Hilla-Merhav** [reply]
Hi @adrianeboyd I hope you're doing well. I was wondering if you had a chance to look into my questions? If you could provide any insights, I would greatly appreciate it. Thank you kindly for your time!

**Reply by adrianeboyd** [reply]
The license for that UD corpus would be okay, but I don't think that the built-in tokenizer in spacy can currently handle the multi-word tokens in the UD Hebrew corpus, and I'm not sure how difficult the lemmatization would be. When I ran UD benchmarks a while ago (https://explosion.ai/blog/ud-benchmarks-v3-2) there wasn't a Hebrew corpus with a commercial-friendly license, so I haven't tried using the trainable edit tree lemmatizer for Hebrew.

Internally, we would only start trying to develop an official spacy trained pipeline (for https://spacy.io/models) if we have all the necessary resources lined up, including the NER corpus.

However, you are always welcome to start developing a pipeline based on the UD corpus or other resources on your own even if it doesn't fit into our trained pipelines. You can potentially start a project similar to https://github.com/huspacy/huspacy or https://github.com/centre-for-humanities-computing/DaCy. We'd be very happy to add a project like this to spaCy universe even if it turns out that an official trained pipeline isn't possible (yet).

One starting point could be the demo project for training a pipeline from a UD corpus:

https://github.com/explosion/projects/tree/v3/pipelines/tagger_parser_ud

Again, the multi-word tokens will be a concern, and spacy just doesn't have a good general solution for this, so tokens would be merged as described in https://explosion.ai/blog/ud-benchmarks-v3-2, which may mean that the Hebrew output doesn't include the kind of annotation that you actually want.

**Reply by Hilla-Merhav** [reply]
Thank you so much for your help and for sharing the good news about the UD corpus license!

Please let me know if I've understood correctly: Is the concept of multi-word tokens related to the following passage in the paper you shared by [Tsarfaty et al. (2019)](https://aclanthology.org/D19-3044.pdf), which describes the SPMRL annotation scheme in Hebrew?

“Due to orthographic and phonological processes, some morphemes do not appear explicitly in the surface form. Our segmentation recovers all morphemes, both overt and covert. For example, the token `בבית` (in the house) is segmented as `׳ב׳, ׳ה׳, ׳בית׳`.”

If this is the problem, this is not an issue in the UD annotation scheme for Hebrew. In contrast to the SPMRL scheme, the UD method employs a different approach that does not involve adding covert morphemes. In the UD scheme, `בבית` is simply segmented into `׳ב׳, ׳בית׳`, and the covert morpheme `ה` (the) is expressed by assigning a specific morphological feature to `ב` (`Definite=Def`).

If there is no other concern, I would like to express my keen interest in developing an official spaCy-trained pipeline for Hebrew. From what I understand, I will need access to both a UD corpus and a NER corpus, but I wanted to inquire if there are any other requirements or resources that I should be aware of in order to successfully execute this project.

I would appreciate your advice on the best tokenizer to use:
1. I am aware that the default tokenizer is regex-based, but I have also read about the new experimental trainable tokenizer and I am curious to learn more about it. 
2. When developing an official spaCy trained pipeline, is it possible to use a third-party, language-specific tokenizer?

**Reply by adrianeboyd** [reply]
I think the main initial concern is the segmentation. With the current built-in components, spacy can only easily support merged multi-word tokens. It's good that the tokenization doesn't insert convert morphemes (the spacy `Doc` object doesn't support this at all), but I still don't think the rule-based tokenizer is going to be able to handle this segmentation.

The experimental trainable tokenizers were mainly developed for the UD benchmark. They're a bit hacky and inefficient, and we're not planning to move them into the core `spacy` library.

For some languages spacy uses third-party segmentation libraries. Are there any existing python Hebrew segmentation libraries that would work for this corpus? Maybe https://github.com/amir-zeldes/RFTokenizer?

The trainable edit tree lemmatizer is a built-in spacy component, so if it works well enough that would be fine.

**Reply by Hilla-Merhav** [reply]
Thank you very much for your help, I am currently searching for the optimal solution for the Hebrew segmentation libraries matter. I am aware that an up-to-date version of the RFTokenizer is available, and I will continue my research on this topic. 
In addition to a tokenizer, are there any other language-specific third-party components that you suggest using, instead of the built-in spacy component? 
You previously mentioned a third-party lemmatizer. Are there any languages supported by spaCy that employ an external lemmatizer?

**Reply by adrianeboyd** [reply]
Russian and Ukrainian use pymorphy3 and Japanese uses sudachipy.

But the built-in edit tree lemmatizer may work fine and it would make sense to test it first before considering third-party libraries.

**Reply by Hilla-Merhav** [reply]
I think I have found a potentially suitable solution for a third-party segmentation library in Hebrew: https://github.com/amir-zeldes/HebPipe
But I would greatly appreciate your advice on it.

**Reply by adrianeboyd** [reply]
HebPipe looks fine for research purposes, but it looks way too heavy as a dependency just for segmentation and we wouldn't want to have a dependency that has such restrictive torch pins because it wouldn't work in practice for many users. Even having spacy suggest a default tokenizer that requires torch is a bit iffy from our side.

The segmentation models look like they may just be `RFTokenizer` models underneath?

I also couldn't easily figure out the details about training corpora and licensing for the provided models?

**Comment by Hilla-Merhav** [comment]
Hello again @adrianeboyd 

I have some questions regarding annotated corpora for NER in Hebrew. It is mentioned [here](https://spacy.io/api/entityrecognizer) that the entity recognizer identifies non-overlapping labelled spans of tokens. I'm a bit unsure about what this implies exactly - does it mean that the entity recognizer doesn't identify nested entities – entity within a longer entity? Should I be concerned if the NER corpus I choose contains nesting?

If you have any additional insights or characteristics that I should consider before selecting a suitable NER corpus, I would greatly appreciate it!

On a related note, based on our previous correspondence, I learned about the importance of avoiding corpora with annotation schemes involving an addition of covert morphemes. Could you please confirm whether this concern is limited to tokenization and UD corpora, or if it also applies to NER corpora? Does it matter if my chosen NER corpus involves adding covert morphemes (as part of the NER annotation)? Thank you for your assistance in clarifying these points.

**Reply by adrianeboyd** [reply]
The main thing is that you need to be able to align the NER annotation to the tokens produced by the tokenizer, which are not going to include covert morphemes. The UD corpus annotation and the NER corpus annotation both need to be able to be aligned to the same tokenization.
